core/services/relay/evm/cap_encoder.go

```diff
@@ -0,0 +1,91 @@
+package evm
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	consensustypes "github.com/smartcontractkit/chainlink-common/pkg/capabilities/consensus/types"
+	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+	abiutil "github.com/smartcontractkit/chainlink/v2/core/chains/evm/abi"
+	"github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/types"
+)
+
+const (
+	abiConfigFieldName = "abi"
+	userEncoderName    = "user"
+	outerEncoderName   = "outer"
+)
+
+// TODO update in common
+const WorkflowIDFieldName = "INTERNAL_worklow_id"
+const WorkflowExecutionIDFieldName = "INTERNAL_workflow_execution_id"
+const InnerUserDataFieldName = "INTERNAL_user_data"
+
+type capEncoder struct {
+	codec commontypes.RemoteCodec
+}
+
+var _ consensustypes.Encoder = (*capEncoder)(nil)
+
+func NewEVMEncoder(config *values.Map) (consensustypes.Encoder, error) {
+	// parse the "inner" encoder config - user-defined fields
+	wrappedSelector, err := config.Underlying[abiConfigFieldName].Unwrap()
+	if err != nil {
+		return nil, err
+	}
+	selectorStr, ok := wrappedSelector.(string)
+	if !ok {
+		return nil, fmt.Errorf("expected %s to be a string", abiConfigFieldName)
+	}
+	selector, err := abiutil.ParseSignature("inner(" + selectorStr + ")")
+	if err != nil {
+		return nil, err
+	}
+	userEncoderConfig, err := json.Marshal(selector.Inputs)
+	if err != nil {
+		return nil, err
+	}
+
+	// prepare the "outer" encoder config - a tuple of (user_fields, workflow_id, workflow_execution_id)
+	outerSelector, err := abiutil.ParseSignature("outer(" +
+		InnerUserDataFieldName + " bytes," +
+		WorkflowIDFieldName + " string," +
+		WorkflowExecutionIDFieldName + " string)")
+	if err != nil {
+		return nil, err
+	}
+	outerEncoderConfig, err := json.Marshal(outerSelector.Inputs)
+	if err != nil {
+		return nil, err
+	}
+
+	codecConfig := types.CodecConfig{Configs: map[string]types.ChainCodecConfig{
+		userEncoderName:  {TypeABI: string(userEncoderConfig)},
+		outerEncoderName: {TypeABI: string(outerEncoderConfig)},
+	}}
+	c, err := NewCodec(codecConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &capEncoder{codec: c}, nil
+}
+
+func (c *capEncoder) Encode(ctx context.Context, input values.Map) ([]byte, error) {
+	unwrappedInput, err := input.Unwrap()
+	if err != nil {
+		return nil, err
+	}
+	unwrappedMap, ok := unwrappedInput.(map[string]any)
+	if !ok {
+		return nil, fmt.Errorf("expected input to be a map")
+	}
```

Comment:
Is this even possible?  You have Map as the input.  I'm not sure I would want to trust the system anymore if that doesn't produce a map[string]any